diff --git a/platform/chromium/vapi-common.js b/platform/chromium/vapi-common.js
index 2b58510d..b2b8b30d 100644
--- a/platform/chromium/vapi-common.js
+++ b/platform/chromium/vapi-common.js
@@ -108,11 +108,6 @@ vAPI.webextFlavor = {
         soup.add('apple').add('safari');
     }
 
-    // https://github.com/gorhill/uBlock/issues/3588
-    if ( soup.has('chromium') && flavor.major >= 66 ) {
-        soup.add('user_stylesheet');
-    }
-
     // Don't starve potential listeners
     vAPI.setTimeout(dispatch, 97);
 })();
diff --git a/platform/chromium/vapi-usercss.js b/platform/chromium/vapi-usercss.js
index 0cf97b7a..cebc2d1d 100644
--- a/platform/chromium/vapi-usercss.js
+++ b/platform/chromium/vapi-usercss.js
@@ -29,8 +29,7 @@
 // https://github.com/gorhill/uBlock/issues/3588
 
 if ( typeof vAPI === 'object' ) {
-    vAPI.supportsUserStylesheets =
-        /\bChrom(?:e|ium)\/(?:5\d|6[012345])\b/.test(navigator.userAgent) === false;
+    vAPI.supportsUserStylesheets = false;
 }
 
 
