diff --git a/src/js/assets.js b/src/js/assets.js
index 4b261a4..fbe4a61 100644
--- a/src/js/assets.js
+++ b/src/js/assets.js
@@ -285,7 +285,7 @@ var registerAssetSource = function(assetKey, dict) {
         entry.contentURL = [];
     }
     if ( typeof entry.updateAfter !== 'number' ) {
-        entry.updateAfter = 5;
+        entry.updateAfter = 3;
     }
     if ( entry.submitter ) {
         entry.submitTime = Date.now(); // To detect stale entries
@@ -904,7 +904,7 @@ api.rmrf = function() {
 // Asset updater area.
 var updaterStatus,
     updaterTimer,
-    updaterAssetDelayDefault = 120000,
+    updaterAssetDelayDefault = 300000,
     updaterAssetDelay = updaterAssetDelayDefault,
     updaterUpdated = [],
     updaterFetched = new Set(),
diff --git a/src/js/background.js b/src/js/background.js
index 2fd1db0..f2856f8 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -40,12 +40,12 @@ var µBlock = (function() { // jshint ignore:line
 
     var hiddenSettingsDefault = {
         assetFetchTimeout: 30,
-        autoUpdateAssetFetchPeriod: 120,
-        autoUpdatePeriod: 7,
+        autoUpdateAssetFetchPeriod: 300,
+        autoUpdatePeriod: 4,
         debugScriptlets: false,
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
-        manualUpdateAssetFetchPeriod: 500,
+        manualUpdateAssetFetchPeriod: 1,
         nanoIgnoreThirdPartyWhitelist: false,
         nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
@@ -58,6 +58,7 @@ var µBlock = (function() { // jshint ignore:line
         'chrome-extension-scheme',
         'chrome-scheme',
         'moz-extension-scheme',
+        'ms-browser-extension-scheme',
         'opera-scheme',
         'vivaldi-scheme',
         'wyciwyg-scheme',   // Firefox's "What-You-Cache-Is-What-You-Get"
diff --git a/src/js/start.js b/src/js/start.js
index 220da63..9d802ee 100644
--- a/src/js/start.js
+++ b/src/js/start.js
@@ -166,13 +166,6 @@ var onUserSettingsReady = function(fetched) {
     µb.permanentURLFiltering.fromString(fetched.urlFilteringString);
     µb.sessionURLFiltering.assign(µb.permanentURLFiltering);
     µb.hnSwitches.fromString(fetched.hostnameSwitchesString);
-
-    // https://github.com/gorhill/uBlock/issues/1892
-    // For first installation on a battery-powered device, disable generic
-    // cosmetic filtering.
-    if ( µb.firstInstall && vAPI.webextFlavor.soup.has('mobile') ) {
-        userSettings.ignoreGenericCosmeticFilters = true;
-    }
 };
 
 /******************************************************************************/
@@ -254,6 +247,7 @@ var onSelectedFilterListsLoaded = function() {
         ].join('\n'),
         'urlFilteringString': '',
         'hostnameSwitchesString': [
+            'no-csp-reports: * true',
             'no-large-media: behind-the-scene false'
         ].join('\n'),
         'lastRestoreFile': '',
diff --git a/src/js/storage.js b/src/js/storage.js
index 8e6d4bd..7fb4da3 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -1275,7 +1275,7 @@
             next = 0;
             var µb = µBlock;
             µb.assets.updateStart({
-                delay: µb.hiddenSettings.autoUpdateAssetFetchPeriod * 1000 || 120000
+                delay: µb.hiddenSettings.autoUpdateAssetFetchPeriod * 1000 || 300000
             });
         }, updateDelay);
     };
@@ -1368,7 +1368,7 @@
             this.loadFilterLists();
         }
         if ( this.userSettings.autoUpdate ) {
-            this.scheduleAssetUpdater(this.hiddenSettings.autoUpdatePeriod * 3600000 || 25200000);
+            this.scheduleAssetUpdater(this.hiddenSettings.autoUpdatePeriod * 3600000 || 10800000);
         } else {
             this.scheduleAssetUpdater(0);
         }
