diff --git a/src/js/assets.js b/src/js/assets.js
index cf1eb98c..9f059e46 100644
--- a/src/js/assets.js
+++ b/src/js/assets.js
@@ -365,7 +365,7 @@ const registerAssetSource = function(assetKey, dict) {
         entry.contentURL = [];
     }
     if ( typeof entry.updateAfter !== 'number' ) {
-        entry.updateAfter = 5;
+        entry.updateAfter = 3;
     }
     if ( entry.submitter ) {
         entry.submitTime = Date.now(); // To detect stale entries
@@ -831,7 +831,7 @@ api.rmrf = function() {
 /******************************************************************************/
 
 // Asset updater area.
-const updaterAssetDelayDefault = 120000;
+const updaterAssetDelayDefault = 180000;
 const updaterUpdated = [];
 const updaterFetched = new Set();
 
diff --git a/src/js/background.js b/src/js/background.js
index 2301c1e1..09b49e76 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -39,9 +39,9 @@ const µBlock = (( ) => { // jshint ignore:line
         allowGenericProceduralFilters: false,
         assetFetchTimeout: 30,
         autoCommentFilterTemplate: '{{date}} {{origin}}',
-        autoUpdateAssetFetchPeriod: 120,
+        autoUpdateAssetFetchPeriod: 180,
         autoUpdateDelayAfterLaunch: 180,
-        autoUpdatePeriod: 7,
+        autoUpdatePeriod: 4,
         benchmarkDatasetURL: 'unset',
         blockingProfiles: '11111/#F00 11011/#C0F 11001/#00F 00001',
         cacheStorageAPI: 'unset',
@@ -59,12 +59,12 @@ const µBlock = (( ) => { // jshint ignore:line
         debugScriptlets: false,
         debugScriptletInjector: false,
         disableWebAssembly: false,
-        extensionUpdateForceReload: false,
+        extensionUpdateForceReload: true,
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         filterAuthorMode: false,
         loggerPopupType: 'popup',
-        manualUpdateAssetFetchPeriod: 500,
+        manualUpdateAssetFetchPeriod: 1,
         nanoIgnoreThirdPartyWhitelist: false,
         nanoMakeThirdPartyFiltersPrivileged: false,
         nanoMakeUserFiltersPrivileged: false,
@@ -91,7 +91,7 @@ const µBlock = (( ) => { // jshint ignore:line
             externalLists: [],
             firewallPaneMinimized: true,
             hyperlinkAuditingDisabled: true,
-            ignoreGenericCosmeticFilters: vAPI.webextFlavor.soup.has('mobile'),
+            ignoreGenericCosmeticFilters: false,
             largeMediaSize: 50,
             nanoEditorSoftWrap: false,
             nanoViewerSoftWrap: true,
@@ -120,8 +120,10 @@ const µBlock = (( ) => { // jshint ignore:line
             'about-scheme',
             'chrome-extension-scheme',
             'chrome-scheme',
+            'edge-extension-scheme',       // For legacy Edge
             'edge-scheme',
             'moz-extension-scheme',
+            'ms-browser-extension-scheme', // For legacy Edge
             'opera-scheme',
             'vivaldi-scheme',
             'wyciwyg-scheme',   // Firefox's "What-You-Cache-Is-What-You-Get"
diff --git a/src/js/start.js b/src/js/start.js
index 04db3da1..1b82e2c5 100644
--- a/src/js/start.js
+++ b/src/js/start.js
@@ -238,6 +238,7 @@ const createDefaultProps = function() {
         ].join('\n'),
         'urlFilteringString': '',
         'hostnameSwitchesString': [
+            'no-csp-reports: * true',
             'no-large-media: behind-the-scene false',
         ].join('\n'),
         'lastRestoreFile': '',
diff --git a/src/js/storage.js b/src/js/storage.js
index e21a9f34..1a820855 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -1305,7 +1305,7 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
             next = 0;
             this.assets.updateStart({
                 delay: this.hiddenSettings.autoUpdateAssetFetchPeriod * 1000 ||
-                       120000
+                       180000
             });
         }, updateDelay);
     };
@@ -1391,7 +1391,7 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
             this.loadFilterLists();
         }
         if ( this.userSettings.autoUpdate ) {
-            this.scheduleAssetUpdater(this.hiddenSettings.autoUpdatePeriod * 3600000 || 25200000);
+            this.scheduleAssetUpdater(this.hiddenSettings.autoUpdatePeriod * 3600000 || 10800000);
         } else {
             this.scheduleAssetUpdater(0);
         }
