diff --git a/src/js/3p-filters.js b/src/js/3p-filters.js
index f34caaa6..7545074d 100644
--- a/src/js/3p-filters.js
+++ b/src/js/3p-filters.js
@@ -496,7 +496,7 @@ const buttonPurgeAllHandler = function(ev) {
         'dashboard',
         {
             what: 'purgeAllCaches',
-            hard: ev.ctrlKey && ev.shiftKey
+            hard: ev.ctrlKey || ev.metaKey || ev.shiftKey
         },
         ( ) => {
             renderFilterLists(true);
