diff --git a/src/js/3p-filters.js b/src/js/3p-filters.js
index 31c80273..6a43cc6f 100644
--- a/src/js/3p-filters.js
+++ b/src/js/3p-filters.js
@@ -670,7 +670,7 @@ uDom('#ignoreGenericCosmeticFilters').on('change', onFilteringSettingsChanged);
 uDom('#buttonApply').on('click', ( ) => { buttonApplyHandler(); });
 uDom('#buttonUpdate').on('click', ( ) => { buttonUpdateHandler(); });
 uDom('#buttonPurgeAll').on('click', ev => {
-    buttonPurgeAllHandler(ev.ctrlKey && ev.shiftKey);
+    buttonPurgeAllHandler(ev.ctrlKey || ev.metaKey || ev.shiftKey);
 });
 uDom('#lists').on('change', '.listEntry > input', onFilteringSettingsChanged);
 uDom('#lists').on('click', '.listEntry > a.remove', onRemoveExternalList);
