diff --git a/src/js/messaging.js b/src/js/messaging.js
index 200c3f6f..7e6e6463 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -949,16 +949,12 @@ const restoreUserData = async function(request) {
     vAPI.app.restart();
 };
 
-// Remove all stored data but keep global counts, people can become
-// quite attached to numbers
 const resetUserData = async function() {
     await Promise.all([
         µb.cacheStorage.clear(),
         vAPI.storage.clear(),
     ]);
 
-    await µb.saveLocalSettings();
-
     vAPI.app.restart();
 };
 
