diff --git a/src/js/messaging.js b/src/js/messaging.js
index b1e0d85..2e257da 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -796,7 +796,7 @@ var restoreUserData = function(request) {
 // quite attached to numbers
 
 var resetUserData = function() {
-    let count = 3;
+    let count = 2;
     let countdown = ( ) => {
         count -= 1;
         if ( count === 0 ) {
@@ -805,8 +805,7 @@ var resetUserData = function() {
     };
     vAPI.cacheStorage.clear(countdown); // 1
     vAPI.storage.clear(countdown);      // 2
-    µb.saveLocalSettings(countdown);    // 3
-    vAPI.localStorage.removeItem('immediateHiddenSettings');
+    vAPI.localStorage.clear();
 };
 
 /******************************************************************************/
