diff --git a/src/js/messaging.js b/src/js/messaging.js
index a112e38c..b2ce50c4 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -957,16 +957,12 @@ const restoreUserData = async function(request) {
     vAPI.app.restart();
 };
 
-// Remove all stored data but keep global counts, people can become
-// quite attached to numbers
 const resetUserData = async function() {
     await Promise.all([
         µb.cacheStorage.clear(),
         vAPI.storage.clear(),
     ]);
 
-    await µb.saveLocalSettings();
-
     vAPI.app.restart();
 };
 
