diff --git a/src/js/messaging.js b/src/js/messaging.js
index 3332c334..3b9fcf48 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -876,11 +876,8 @@ const restoreUserData = function(request) {
     vAPI.localStorage.removeItem('immediateHiddenSettings');
 };
 
-// Remove all stored data but keep global counts, people can become
-// quite attached to numbers
-
 const resetUserData = function() {
-    let count = 3;
+    let count = 2;
     const countdown = ( ) => {
         count -= 1;
         if ( count === 0 ) {
@@ -889,8 +886,7 @@ const resetUserData = function() {
     };
     µb.cacheStorage.clear().then(( ) => countdown());   // 1
     vAPI.storage.clear(countdown);                      // 2
-    µb.saveLocalSettings(countdown);                    // 3
-    vAPI.localStorage.removeItem('immediateHiddenSettings');
+    vAPI.localStorage.clear();
 };
 
 /******************************************************************************/
