diff --git a/src/js/messaging.js b/src/js/messaging.js
index 4396f3d6..3b9d3976 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -938,18 +938,14 @@ const restoreUserData = async function(request) {
     vAPI.app.restart();
 };
 
-// Remove all stored data but keep global counts, people can become
-// quite attached to numbers
 const resetUserData = async function() {
-    vAPI.localStorage.removeItem('immediateHiddenSettings');
+    vAPI.localStorage.clear();
 
     await Promise.all([
         µb.cacheStorage.clear(),
         vAPI.storage.clear(),
     ]);
 
-    await µb.saveLocalSettings();
-
     vAPI.app.restart();
 };
 
