diff --git a/src/js/messaging.js b/src/js/messaging.js
index c24c592b..c02b953e 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -849,11 +849,8 @@ var restoreUserData = function(request) {
     vAPI.localStorage.removeItem('immediateHiddenSettings');
 };
 
-// Remove all stored data but keep global counts, people can become
-// quite attached to numbers
-
 var resetUserData = function() {
-    let count = 3;
+    let count = 2;
     let countdown = ( ) => {
         count -= 1;
         if ( count === 0 ) {
@@ -862,8 +859,7 @@ var resetUserData = function() {
     };
     µb.cacheStorage.clear(countdown); // 1
     vAPI.storage.clear(countdown);    // 2
-    µb.saveLocalSettings(countdown);  // 3
-    vAPI.localStorage.removeItem('immediateHiddenSettings');
+    vAPI.localStorage.clear();
 };
 
 /******************************************************************************/
