diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index e08d5b04..e2d95a0a 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -195,7 +195,9 @@
             content =
                 'try {\n' +
                     content.replace('{{nano}}', '') + '\n' +
-                '} catch ( e ) { }';
+                '} catch ( e ) {\n' +
+                '    console.error("[Nano] Script Snippet ::", e);\n' +
+                '}';
             scriptletCache.add(raw, content);
         }
         toInject.set(raw, content);
