diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index eab9ff9b..87e54fa7 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -230,7 +230,9 @@
             content =
                 'try {\n' +
                     content.replace('{{nano}}', '') + '\n' +
-                '} catch ( e ) { }';
+                '} catch ( e ) {\n' +
+                '    console.error("[Nano] Script Snippet ::", e);\n' +
+                '}';
             scriptletCache.add(rawToken, content);
         }
         toInject.set(rawToken, content);
