diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index 0ac39cea..5bfd6aba 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -187,7 +187,9 @@
             content =
                 'try {\n' +
                     content.replace('{{nano}}', '') + '\n' +
-                '} catch ( e ) { }';
+                '} catch ( e ) {\n' +
+                '    console.error("[Nano] Script Snippet ::", e);\n' +
+                '}';
             scriptletCache.add(raw, content);
         }
         toInject.set(raw, content);
