diff --git a/src/js/scriptlets/element-picker.js b/src/js/scriptlets/element-picker.js
index 1cb5ecb2..e950a019 100644
--- a/src/js/scriptlets/element-picker.js
+++ b/src/js/scriptlets/element-picker.js
@@ -1153,6 +1153,7 @@ const onDialogClicked = function(ev) {
                     what: 'createUserFilter',
                     autoComment: true,
                     filters: filter,
+                    nanoHref: window.location.href,
                     origin: window.location.origin,
                     pageDomain: window.location.hostname
                 }
diff --git a/src/js/storage.js b/src/js/storage.js
index 16335408..27cc4475 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -437,7 +437,8 @@
             this.hiddenSettings.autoCommentFilterTemplate
                 .replace('{{date}}', d.toLocaleDateString())
                 .replace('{{time}}', d.toLocaleTimeString())
-                .replace('{{origin}}', options.origin);
+                .replace('{{origin}}', options.origin)
+                .replace('{{nanoHref}}', options.nanoHref);
     }
 
     const onSaved = ( ) => {
