diff --git a/src/js/scriptlets/element-picker.js b/src/js/scriptlets/element-picker.js
index 1a366c39..cb9502d7 100644
--- a/src/js/scriptlets/element-picker.js
+++ b/src/js/scriptlets/element-picker.js
@@ -1186,6 +1186,7 @@ const onDialogClicked = function(ev) {
                 what: 'createUserFilter',
                 autoComment: true,
                 filters: filter,
+                nanoHref: window.location.href,
                 origin: window.location.origin,
                 pageDomain: window.location.hostname,
                 killCache: isCosmetic === false,
diff --git a/src/js/storage.js b/src/js/storage.js
index 4a76f5d9..79e3da9f 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -422,7 +422,8 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
             this.hiddenSettings.autoCommentFilterTemplate
                 .replace('{{date}}', d.toLocaleDateString())
                 .replace('{{time}}', d.toLocaleTimeString())
-                .replace('{{origin}}', options.origin);
+                .replace('{{origin}}', options.origin)
+                .replace('{{nanoHref}}', options.nanoHref);
     }
 
     const details = await this.loadUserFilters();
