diff --git a/src/js/scriptlets/element-picker.js b/src/js/scriptlets/element-picker.js
index 5f3e6ee8..676f43c5 100644
--- a/src/js/scriptlets/element-picker.js
+++ b/src/js/scriptlets/element-picker.js
@@ -1186,6 +1186,7 @@ const onDialogClicked = function(ev) {
                 what: 'createUserFilter',
                 autoComment: true,
                 filters: filter,
+                nanoHref: window.location.href,
                 origin: window.location.origin,
                 pageDomain: window.location.hostname,
                 killCache: isCosmetic === false,
diff --git a/src/js/storage.js b/src/js/storage.js
index a650a169..3f7822bd 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -411,7 +411,8 @@
             this.hiddenSettings.autoCommentFilterTemplate
                 .replace('{{date}}', d.toLocaleDateString())
                 .replace('{{time}}', d.toLocaleTimeString())
-                .replace('{{origin}}', options.origin);
+                .replace('{{origin}}', options.origin)
+                .replace('{{nanoHref}}', options.nanoHref);
     }
 
     const details = await this.loadUserFilters();
