diff --git a/src/js/3p-filters.js b/src/js/3p-filters.js
index 42ea241f..f81090b9 100644
--- a/src/js/3p-filters.js
+++ b/src/js/3p-filters.js
@@ -25,6 +25,10 @@
 
 /******************************************************************************/
 
+window.nano_has_unsaved_changes = false;
+
+/******************************************************************************/
+
 (function() {
 
 /******************************************************************************/
@@ -309,9 +313,11 @@ var renderFilterLists = function(soft) {
 /******************************************************************************/
 
 var renderWidgets = function() {
+    var nano_is_clean = filteringSettingsHash === hashFromCurrentFromSettings();
+    nano_has_unsaved_changes = !nano_is_clean;
     uDom('#buttonApply').toggleClass(
         'disabled',
-        filteringSettingsHash === hashFromCurrentFromSettings()
+        nano_is_clean
     );
     uDom('#buttonPurgeAll').toggleClass(
         'disabled',
diff --git a/src/js/dashboard.js b/src/js/dashboard.js
index d038c5d0..5ec05c8b 100644
--- a/src/js/dashboard.js
+++ b/src/js/dashboard.js
@@ -62,7 +62,10 @@ let onTabClickHandler = function(e) {
         url = url.slice(0, pos);
     }
     url += this.hash;
-    if ( url !== window.location.href ) {
+    if (
+        url !== window.location.href &&
+        ( !nano_check_unsaved_changes() || confirm(nano_unsaved_warning_msg) )
+    ) {
         window.location.replace(url);
         loadDashboardPanel();
     }
@@ -81,4 +84,32 @@ loadDashboardPanel();
 
 /******************************************************************************/
 
+let nano_content_iframe = document.getElementById('iframe');
+let nano_unsaved_warning_msg = vAPI.i18n('nano_d_unsaved_warning');
+
+let nano_check_unsaved_changes = function() {
+    const iframe = nano_content_iframe;
+    if ( iframe.src == '' ) {
+        return false;
+    } else {
+        return iframe.contentWindow.nano_has_unsaved_changes === true;
+    }
+};
+
+addEventListener('beforeunload', function(e) {
+    if ( nano_check_unsaved_changes() ) {
+        const msg = nano_check_unsaved_changes.msg;
+        // Old standard
+        e.returnValue = msg;
+        // New standard
+        if ( typeof e.preventDefault === 'function' ) {
+            e.preventDefault();
+        }
+        // Just in case
+        return msg;
+    }
+});
+
+/******************************************************************************/
+
 })();
diff --git a/src/js/dyna-rules.js b/src/js/dyna-rules.js
index 5e36ce91..cf4fc938 100644
--- a/src/js/dyna-rules.js
+++ b/src/js/dyna-rules.js
@@ -25,6 +25,10 @@
 
 /******************************************************************************/
 
+window.nano_has_unsaved_changes = false;
+
+/******************************************************************************/
+
 (function() {
 
 /******************************************************************************/
@@ -385,6 +389,7 @@ const onTextChanged = (function() {
             input.setAttribute('disabled', '');
             CodeMirror.commands.save = editSaveHandler;
         }
+        nano_has_unsaved_changes = !isClean;
     };
 
     return function(now) {
diff --git a/src/js/udom.js b/src/js/udom.js
index dc85e6b2..553b4b34 100644
--- a/src/js/udom.js
+++ b/src/js/udom.js
@@ -675,20 +675,6 @@ DOMList.prototype.trigger = function(etype) {
 
 /******************************************************************************/
 
-// Cleanup
-
-var onBeforeUnload = function() {
-    var entry;
-    while ( (entry = listenerEntries.pop()) ) {
-        entry.dispose();
-    }
-    window.removeEventListener('beforeunload', onBeforeUnload);
-};
-
-window.addEventListener('beforeunload', onBeforeUnload);
-
-/******************************************************************************/
-
 return DOMListFactory;
 
 })();
diff --git a/src/js/whitelist.js b/src/js/whitelist.js
index 13456f01..dd6980c9 100644
--- a/src/js/whitelist.js
+++ b/src/js/whitelist.js
@@ -25,6 +25,10 @@
 
 /******************************************************************************/
 
+window.nano_has_unsaved_changes = false;
+
+/******************************************************************************/
+
 (function() {
 
 /******************************************************************************/
@@ -90,6 +94,7 @@ var whitelistChanged = function() {
     uDom.nodeFromId('whitelistApply').disabled = !changed || bad;
     uDom.nodeFromId('whitelistRevert').disabled = !changed;
     CodeMirror.commands.save = changed && !bad ? applyChanges : noopFunc;
+    nano_has_unsaved_changes = changed;
 };
 
 cmEditor.on('changes', whitelistChanged);
