diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index 5e009197..f2f67e83 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -731,6 +731,7 @@ const removeTopCommentBlock = function(text) {
 RedirectEngine.prototype.loadBuiltinResources = function() {
     // TODO: remove once usage of uBO 1.20.4 is widespread.
     µBlock.assets.remove('ublock-resources');
+    µBlock.assets.remove('nano-resources');
 
     this.resources = new Map();
     this.aliases = new Map();
