diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index c5eab189..6c08209c 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -711,6 +711,7 @@ RedirectEngine.prototype.loadBuiltinResources = function() {
 
     // TODO: remove once usage of uBO 1.20.4 is widespread.
     µBlock.assets.remove('ublock-resources');
+    µBlock.assets.remove('nano-resources');
 
     for ( const [ name, details ] of redirectableResources ) {
         if ( details.inject !== false ) {
