diff --git a/src/js/background.js b/src/js/background.js
index 364468fa..ed23581c 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -143,8 +143,8 @@ const µBlock = (function() { // jshint ignore:line
 
         // Read-only
         systemSettings: {
-            compiledMagic: 14,  // Increase when compiled format changes
-            selfieMagic: 14     // Increase when selfie format changes
+            compiledMagic: 5,  // Increase when compiled format changes
+            selfieMagic: 5     // Increase when selfie format changes
         },
 
         restoreBackupSettings: {
