diff --git a/platform/chromium/vapi-background.js b/platform/chromium/vapi-background.js
index 461a8e79..fd5cb70c 100644
--- a/platform/chromium/vapi-background.js
+++ b/platform/chromium/vapi-background.js
@@ -57,16 +57,8 @@ window.addEventListener('webextFlavor', function() {
 /******************************************************************************/
 
 vAPI.app = {
-    name: manifest.name.replace(/ dev\w+ build/, ''),
-    version: (( ) => {
-        let version = manifest.version;
-        const match = /(\d+\.\d+\.\d+)(?:\.(\d+))?/.exec(version);
-        if ( match && match[2] ) {
-            const v = parseInt(match[2], 10);
-            version = match[1] + (v < 100 ? 'b' + v : 'rc' + (v - 100));
-        }
-        return version;
-    })(),
+    name: manifest.name,
+    version: manifest.version,
 
     intFromVersion: function(s) {
         const parts = s.match(/(?:^|\.|b|rc)\d+/g);
