diff --git a/platform/chromium/vapi-background.js b/platform/chromium/vapi-background.js
index b876089b..9f5359bc 100644
--- a/platform/chromium/vapi-background.js
+++ b/platform/chromium/vapi-background.js
@@ -662,10 +662,10 @@ vAPI.setIcon = (function() {
             ' ({badge})';
     const icons = [
         {
-            path: { '16': 'img/icon_16-off.png', '32': 'img/icon_32-off.png' }
+            path: { '128': 'img/128_off.png' }
         },
         {
-            path: { '16': 'img/icon_16.png', '32': 'img/icon_32.png' }
+            path: { '128': 'img/128_on.png' }
         }
     ];
 
