diff --git a/src/js/background.js b/src/js/background.js
index b3e761fa..d99a9be5 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -53,6 +53,9 @@ const µBlock = (function() { // jshint ignore:line
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeThirdPartyFiltersPrivileged: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         requestJournalProcessPeriod: 1000,
         selfieAfter: 3,
@@ -159,6 +162,7 @@ const µBlock = (function() { // jshint ignore:line
         assetsBootstrapLocation: undefined,
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index 6e4a6d1d..f8adb9c1 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -800,6 +800,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 04e2fa3b..4d92d478 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -2028,6 +2028,10 @@ FilterParser.prototype.parseOptions = function(s) {
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
         if ( opt === 'badfilter' ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                break;
+            }
             this.badFilter = true;
             continue;
         }
@@ -2111,6 +2115,10 @@ FilterParser.prototype.parse = function(raw) {
     // block or allow filter?
     // Important: this must be executed before parsing options
     if ( s.startsWith('@@') ) {
+        if ( nano.cf.strip_whitelist ) {
+            this.unsupported = true;
+            return this;
+        }
         this.action = AllowAction;
         s = s.slice(2);
     }
diff --git a/src/js/storage.js b/src/js/storage.js
index 2d1619b7..f57d928f 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -448,7 +448,8 @@
     const onSaved = ( ) => {
         const compiledFilters = this.compileFilters(
             filters,
-            { assetKey: this.userFiltersPath }
+            { assetKey: this.userFiltersPath },
+            nano.ub.nanoPartialUserFiltersPath
         );
         const snfe = this.staticNetFilteringEngine;
         const cfe = this.cosmeticFilteringEngine;
@@ -846,7 +847,7 @@
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText, details) {
+µBlock.compileFilters = function(rawText, details, nanoAssetKeyOverride) {
     let writer = new this.CompiledLineIO.Writer();
 
     // Populate the writer with information potentially useful to the
@@ -857,6 +858,9 @@
         }
     }
 
+    const nanoKey = nanoAssetKeyOverride || details.assetKey;
+    nano.cf.update(nanoKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -916,6 +920,8 @@
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
