diff --git a/src/js/background.js b/src/js/background.js
index 2cf32f60..b3055cf3 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -50,6 +50,9 @@ const µBlock = (function() { // jshint ignore:line
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeThirdPartyFiltersPrivileged: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         requestJournalProcessPeriod: 1000,
         strictBlockingBypassDuration: 120,
@@ -155,6 +157,7 @@ const µBlock = (function() { // jshint ignore:line
         assetsBootstrapLocation: 'assets/assets.json',
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index 6a0b1deb..b2ca3612 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -709,6 +709,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 99003116..197eebd6 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1561,6 +1561,10 @@ FilterParser.prototype.parseOptions = function(s) {
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
         if ( opt === 'badfilter' ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                break;
+            }
             this.badFilter = true;
             continue;
         }
@@ -1700,6 +1704,10 @@ FilterParser.prototype.parse = function(raw) {
     // block or allow filter?
     // Important: this must be executed before parsing options
     if ( s.startsWith('@@') ) {
+        if ( nano.cf.strip_whitelist ) {
+            this.unsupported = true;
+            return this;
+        }
         this.action = AllowAction;
         s = s.slice(2);
     }
diff --git a/src/js/storage.js b/src/js/storage.js
index 05dc0e72..f7b3d82a 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -410,7 +410,7 @@
     var µb = this;
 
     var onSaved = function() {
-        var compiledFilters = µb.compileFilters(filters),
+        var compiledFilters = µb.compileFilters(filters, µb.nanoPartialUserFiltersPath),
             snfe = µb.staticNetFilteringEngine,
             cfe = µb.cosmeticFilteringEngine,
             acceptedCount = snfe.acceptedCount + cfe.acceptedCount,
@@ -702,7 +702,7 @@
 
     var onCompiledListLoaded2 = function(details) {
         if ( details.content === '' ) {
-            details.content = µb.compileFilters(rawContent);
+            details.content = µb.compileFilters(rawContent, assetKey);
             µb.assets.put(compiledPath, details.content);
         }
         rawContent = undefined;
@@ -784,9 +784,11 @@
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText) {
+µBlock.compileFilters = function(rawText, nanoKey) {
     let writer = new this.CompiledLineIO.Writer();
 
+    nano.cf.update(nanoKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -846,6 +848,8 @@
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
@@ -1256,7 +1260,7 @@
                     );
                     this.assets.put(
                         'compiled/' + details.assetKey,
-                        this.compileFilters(details.content)
+                        this.compileFilters(details.content, details.assetKey)
                     );
                 }
             } else {
