diff --git a/src/js/background.js b/src/js/background.js
index b724336f..d8c309ad 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -50,6 +50,9 @@ const µBlock = (function() { // jshint ignore:line
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeThirdPartyFiltersPrivileged: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         requestJournalProcessPeriod: 1000,
         strictBlockingBypassDuration: 120,
@@ -155,6 +158,7 @@ const µBlock = (function() { // jshint ignore:line
         assetsBootstrapLocation: 'assets/assets.json',
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index 87101558..1208b88a 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -749,6 +749,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index dc8382d4..f9c91647 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1586,6 +1586,10 @@ FilterParser.prototype.parseOptions = function(s) {
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
         if ( opt === 'badfilter' ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                break;
+            }
             this.badFilter = true;
             continue;
         }
@@ -1725,6 +1729,10 @@ FilterParser.prototype.parse = function(raw) {
     // block or allow filter?
     // Important: this must be executed before parsing options
     if ( s.startsWith('@@') ) {
+        if ( nano.cf.strip_whitelist ) {
+            this.unsupported = true;
+            return this;
+        }
         this.action = AllowAction;
         s = s.slice(2);
     }
diff --git a/src/js/storage.js b/src/js/storage.js
index 9bc87342..ae385e46 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -410,7 +410,8 @@
     const onSaved = ( ) => {
         const compiledFilters = this.compileFilters(
             filters,
-            { assetKey: this.userFiltersPath }
+            { assetKey: this.userFiltersPath },
+            µb.nanoPartialUserFiltersPath
         );
         const snfe = this.staticNetFilteringEngine;
         const cfe = this.cosmeticFilteringEngine;
@@ -794,7 +795,7 @@
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText, details) {
+µBlock.compileFilters = function(rawText, details, nanoAssetKeyOverride) {
     let writer = new this.CompiledLineIO.Writer();
 
     // Populate the writer with information potentially useful to the
@@ -805,6 +806,8 @@
         }
     }
 
+    nano.cf.update(nanoAssetKeyOverride || details.assetKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -864,6 +867,8 @@
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
