diff --git a/src/js/background.js b/src/js/background.js
index d711a615..e70090a0 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -46,6 +46,8 @@ var µBlock = (function() { // jshint ignore:line
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         suspendTabsUntilReady: false,
         userResourcesLocation: 'unset'
@@ -156,6 +158,7 @@ var µBlock = (function() { // jshint ignore:line
         assetsBootstrapLocation: 'assets/assets.json',
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index dea5ad30..5161a9d9 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -644,6 +644,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 5849efa0..a49ef008 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1594,6 +1594,10 @@ FilterParser.prototype.parseOptions = function(s) {
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
         if ( opt === 'badfilter' ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                break;
+            }
             this.badFilter = BadFilter;
             continue;
         }
@@ -1733,6 +1737,10 @@ FilterParser.prototype.parse = function(raw) {
     // block or allow filter?
     // Important: this must be executed before parsing options
     if ( s.startsWith('@@') ) {
+        if ( nano.cf.strip_whitelist ) {
+            this.unsupported = true;
+            return this;
+        }
         this.action = AllowAction;
         s = s.slice(2);
     }
diff --git a/src/js/storage.js b/src/js/storage.js
index f8c429db..856abab3 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -411,7 +411,7 @@
     var µb = this;
 
     var onSaved = function() {
-        var compiledFilters = µb.compileFilters(filters),
+        var compiledFilters = µb.compileFilters(filters, µb.nanoPartialUserFiltersPath),
             snfe = µb.staticNetFilteringEngine,
             cfe = µb.cosmeticFilteringEngine,
             acceptedCount = snfe.acceptedCount + cfe.acceptedCount,
@@ -705,7 +705,7 @@
 
     var onCompiledListLoaded2 = function(details) {
         if ( details.content === '' ) {
-            details.content = µb.compileFilters(rawContent);
+            details.content = µb.compileFilters(rawContent, assetKey);
             µb.assets.put(compiledPath, details.content);
         }
         rawContent = undefined;
@@ -788,9 +788,11 @@
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText) {
+µBlock.compileFilters = function(rawText, nanoKey) {
     var writer = new this.CompiledLineWriter();
 
+    nano.cf.update(nanoKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -851,6 +853,8 @@
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
@@ -1259,7 +1263,7 @@
                     );
                     this.assets.put(
                         'compiled/' + details.assetKey,
-                        this.compileFilters(details.content)
+                        this.compileFilters(details.content, details.assetKey)
                     );
                 }
             } else {
