diff --git a/src/js/background.js b/src/js/background.js
index 39439363..78d42308 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -49,6 +49,9 @@ const µBlock = (function() { // jshint ignore:line
         ignoreRedirectFilters: false,
         ignoreScriptInjectFilters: false,
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeThirdPartyFiltersPrivileged: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         requestJournalProcessPeriod: 1000,
         selfieAfter: 11,
@@ -156,6 +159,7 @@ const µBlock = (function() { // jshint ignore:line
         assetsBootstrapLocation: 'assets/assets.json',
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index e805777d..4c41f7b1 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -802,6 +802,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index b71be872..be1211a0 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1963,6 +1963,10 @@ FilterParser.prototype.parseOptions = function(s) {
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
         if ( opt === 'badfilter' ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                break;
+            }
             this.badFilter = true;
             continue;
         }
@@ -2012,6 +2016,10 @@ FilterParser.prototype.parse = function(raw) {
     // block or allow filter?
     // Important: this must be executed before parsing options
     if ( s.startsWith('@@') ) {
+        if ( nano.cf.strip_whitelist ) {
+            this.unsupported = true;
+            return this;
+        }
         this.action = AllowAction;
         s = s.slice(2);
     }
diff --git a/src/js/storage.js b/src/js/storage.js
index a2ce93e8..2f4caeb4 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -449,7 +449,8 @@
     const onSaved = ( ) => {
         const compiledFilters = this.compileFilters(
             filters,
-            { assetKey: this.userFiltersPath }
+            { assetKey: this.userFiltersPath },
+            nano.ub.nanoPartialUserFiltersPath
         );
         const snfe = this.staticNetFilteringEngine;
         const cfe = this.cosmeticFilteringEngine;
@@ -844,7 +845,7 @@
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText, details) {
+µBlock.compileFilters = function(rawText, details, nanoAssetKeyOverride) {
     let writer = new this.CompiledLineIO.Writer();
 
     // Populate the writer with information potentially useful to the
@@ -855,6 +856,9 @@
         }
     }
 
+    const nanoKey = nanoAssetKeyOverride || details.assetKey;
+    nano.cf.update(nanoKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -914,6 +918,8 @@
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
