diff --git a/src/js/background.js b/src/js/background.js
index 2641c3dd..cbe9c0fc 100644
--- a/src/js/background.js
+++ b/src/js/background.js
@@ -63,6 +63,9 @@ const µBlock = (( ) => { // jshint ignore:line
         filterAuthorMode: false,
         loggerPopupType: 'popup',
         manualUpdateAssetFetchPeriod: 500,
+        nanoIgnoreThirdPartyWhitelist: false,
+        nanoMakeThirdPartyFiltersPrivileged: false,
+        nanoMakeUserFiltersPrivileged: false,
         popupFontSize: 'unset',
         requestJournalProcessPeriod: 1000,
         selfieAfter: 3,
@@ -178,6 +181,7 @@ const µBlock = (( ) => { // jshint ignore:line
         assetsBootstrapLocation: undefined,
 
         userFiltersPath: 'user-filters',
+        nanoPartialUserFiltersPath: 'nano-partial-user-filters',
         pslAssetKey: 'public_suffix_list.dat',
 
         selectedFilterLists: [],
diff --git a/src/js/static-ext-filtering.js b/src/js/static-ext-filtering.js
index 1b0be890..f4448e62 100644
--- a/src/js/static-ext-filtering.js
+++ b/src/js/static-ext-filtering.js
@@ -855,6 +855,8 @@
         // Exception filter?
         parsed.exception = raw.charCodeAt(lpos + 1) === 0x40 /* '@' */;
 
+        if ( parsed.exception && nano.cf.strip_whitelist ) { return true; }
+
         // Extract the hostname(s), punycode if required.
         if ( lpos === 0 ) {
             parsed.hostnames = emptyArray;
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 0fd0908a..5b8a8fca 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -2288,6 +2288,10 @@ const FilterParser = class {
             }
             // https://github.com/uBlockOrigin/uAssets/issues/192
             if ( opt === 'badfilter' ) {
+                if ( nano.cf.strip_whitelist ) {
+                    this.unsupported = true;
+                    break;
+                }
                 this.badFilter = true;
                 continue;
             }
@@ -2373,6 +2377,10 @@ const FilterParser = class {
         // block or allow filter?
         // Important: this must be executed before parsing options
         if ( s.startsWith('@@') ) {
+            if ( nano.cf.strip_whitelist ) {
+                this.unsupported = true;
+                return this;
+            }
             this.action = AllowAction;
             s = s.slice(2);
         }
diff --git a/src/js/storage.js b/src/js/storage.js
index a3994b5c..0a954581 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -444,7 +444,8 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
 
     const compiledFilters = this.compileFilters(
         filters,
-        { assetKey: this.userFiltersPath }
+        { assetKey: this.userFiltersPath },
+        nano.ub.nanoPartialUserFiltersPath
     );
     const snfe = this.staticNetFilteringEngine;
     const cfe = this.cosmeticFilteringEngine;
@@ -809,7 +810,7 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
 
 /******************************************************************************/
 
-µBlock.compileFilters = function(rawText, details) {
+µBlock.compileFilters = function(rawText, details, nanoAssetKeyOverride) {
     const writer = new this.CompiledLineIO.Writer();
 
     // Populate the writer with information potentially useful to the
@@ -820,6 +821,9 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
         }
     }
 
+    const nanoKey = nanoAssetKeyOverride || details.assetKey;
+    nano.cf.update(nanoKey);
+
     // Useful references:
     //    https://adblockplus.org/en/filter-cheatsheet
     //    https://adblockplus.org/en/filters
@@ -879,6 +883,8 @@ self.addEventListener('hiddenSettingsChanged', ( ) => {
         staticNetFilteringEngine.compile(line, writer);
     }
 
+    nano.cf.reset();
+
     return writer.toString();
 };
 
