diff --git a/src/js/assets.js b/src/js/assets.js
index 8e1c0dbb..b857cde8 100644
--- a/src/js/assets.js
+++ b/src/js/assets.js
@@ -988,6 +988,9 @@ const updateNext = function() {
             if ( assetKey === 'ublock-resources' && noRemoteResources ) {
                 continue;
             }
+            if ( assetKey === 'nano-resources' && noRemoteResources ) {
+                continue;
+            }
             if (
                 fireNotification(
                     'before-asset-updated',
diff --git a/src/js/messaging.js b/src/js/messaging.js
index b824924f..4b3b1b5c 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1083,10 +1083,14 @@ const onMessage = function(request, sender, callback) {
         µb.assets.purge(request.assetKey);
         µb.assets.remove('compiled/' + request.assetKey);
         // https://github.com/gorhill/uBlock/pull/2314#issuecomment-278716960
-        if ( request.assetKey === 'ublock-filters' ) {
+        if ( request.assetKey.startsWith('ublock-') ) {
             µb.assets.purge('ublock-resources');
             µb.redirectEngine.invalidateResourcesSelfie();
         }
+        if ( request.assetKey.startsWith('nano-') ) {
+            µb.assets.purge('nano-resources');
+            µb.redirectEngine.invalidateResourcesSelfie();
+        }
         break;
 
     case 'readHiddenSettings':
diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index 4f9ddee4..e370611c 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -194,7 +194,7 @@
             }
             content =
                 'try {\n' +
-                    content + '\n' +
+                    content.replace('{{nano}}', '') + '\n' +
                 '} catch ( e ) { }';
             scriptletCache.add(raw, content);
         }
@@ -257,6 +257,14 @@
             return;
         }
 
+        var nanoInjectArgs = parsed.suffix.slice(4, -1);
+        if (
+            !nano.cf.is_privileged &&
+            nanoInjectArgs.startsWith(nano.privileged_assets_previx)
+        ) {
+            return;
+        }
+
         // https://github.com/gorhill/uBlock/issues/3375
         //   Ignore instances of exception filter with negated hostnames,
         //   because there is no way to create an exception to an exception.
diff --git a/src/js/storage.js b/src/js/storage.js
index ebf47b10..f916d1da 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -992,7 +992,10 @@
     return this.redirectEngine.resourcesFromSelfie().then(success => {
         if ( success === true ) { return; }
 
-        const fetchPromises = [ this.assets.get('ublock-resources') ];
+        const fetchPromises = [ 
+            this.assets.get('ublock-resources'),
+            this.assets.get('nano-resources')
+        ];
 
         const userResourcesLocation = this.hiddenSettings.userResourcesLocation;
         if ( userResourcesLocation !== 'unset' ) {
@@ -1310,7 +1313,10 @@
             }
         }
         // https://github.com/gorhill/uBlock/issues/2594
-        if ( details.assetKey === 'ublock-resources' ) {
+        if (
+                details.assetKey === 'ublock-resources' ||
+                details.assetKey === 'nano-resources'
+        ) {
             if (
                 this.hiddenSettings.ignoreRedirectFilters === true &&
                 this.hiddenSettings.ignoreScriptInjectFilters === true
@@ -1349,7 +1355,10 @@
             if ( cached ) {
                 this.compilePublicSuffixList(details.content);
             }
-        } else if ( details.assetKey === 'ublock-resources' ) {
+        } else if (
+            details.assetKey === 'ublock-resources' ||
+            details.assetKey === 'nano-resources'
+        ) {
             this.redirectEngine.invalidateResourcesSelfie();
         }
         vAPI.messaging.broadcast({
