diff --git a/src/js/assets.js b/src/js/assets.js
index f1da642a..b6821bbd 100644
--- a/src/js/assets.js
+++ b/src/js/assets.js
@@ -954,6 +954,9 @@ var updateNext = function() {
             if ( assetKey === 'ublock-resources' && noRemoteResources ) {
                 continue;
             }
+            if ( assetKey === 'nano-resources' && noRemoteResources ) {
+                continue;
+            }
             if (
                 fireNotification(
                     'before-asset-updated',
diff --git a/src/js/messaging.js b/src/js/messaging.js
index 4d64f42f..94e0d6ff 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1074,10 +1074,14 @@ var onMessage = function(request, sender, callback) {
         µb.assets.purge(request.assetKey);
         µb.assets.remove('compiled/' + request.assetKey);
         // https://github.com/gorhill/uBlock/pull/2314#issuecomment-278716960
-        if ( request.assetKey === 'ublock-filters' ) {
+        if ( request.assetKey.startsWith('ublock-') ) {
             µb.assets.purge('ublock-resources');
             µb.redirectEngine.invalidateResourcesSelfie();
         }
+        if ( request.assetKey.startsWith('nano-') ) {
+            µb.assets.purge('nano-resources');
+            µb.redirectEngine.invalidateResourcesSelfie();
+        }
         break;
 
     case 'readHiddenSettings':
diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index 258d580e..ea3bb631 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -186,7 +186,7 @@
             }
             content =
                 'try {\n' +
-                    content + '\n' +
+                    content.replace('{{nano}}', '') + '\n' +
                 '} catch ( e ) { }';
             scriptletCache.add(raw, content);
         }
@@ -249,6 +249,14 @@
             return;
         }
 
+        var nanoInjectArgs = parsed.suffix.slice(4, -1);
+        if (
+            !nano.cf.is_privileged &&
+            nanoInjectArgs.startsWith(nano.privileged_assets_previx)
+        ) {
+            return;
+        }
+
         // https://github.com/gorhill/uBlock/issues/3375
         //   Ignore instances of exception filter with negated hostnames,
         //   because there is no way to create an exception to an exception.
diff --git a/src/js/storage.js b/src/js/storage.js
index f7b3d82a..2c8f73f8 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -924,7 +924,7 @@
 
 /******************************************************************************/
 
-µBlock.loadRedirectResources = function(updatedContent) {
+µBlock.loadRedirectResources = function(updatedContent, nanoIsExtended) {
     var µb = this,
         content = '';
 
@@ -939,9 +939,28 @@
         onDone();
     };
 
-    var onResourcesLoaded = function(details) {
+    var nanoHasChange =
+        typeof updatedContent === 'string' && updatedContent.length !== 0;
+
+    var nanoFetchResourceByKey = function(key, callback) {
+        if ( !nanoHasChange ) {
+            µb.assets.get(key, callback);
+            return;
+        }
+        if ( key === 'ublock-resources' && !nanoIsExtended ) {
+            callback({ content: updatedContent });
+            return;
+        }
+        if ( key === 'nano-resources' && nanoIsExtended ) {
+            callback({ content: updatedContent });
+            return;
+        }
+        µb.assets.get(key, callback);
+    };
+
+    var nanoOnExtendedResourcesLoaded = function(details) {
         if ( details.content !== '' ) {
-            content = details.content;
+            content += '\n\n' + details.content;
         }
         if ( µb.hiddenSettings.userResourcesLocation === 'unset' ) {
             return onDone();
@@ -949,16 +968,24 @@
         µb.assets.fetchText(µb.hiddenSettings.userResourcesLocation, onUserResourcesLoaded);
     };
 
-    if ( typeof updatedContent === 'string' && updatedContent.length !== 0 ) {
-        return onResourcesLoaded({ content: updatedContent });
-    }
+    var nanoOnResourcesLoaded = function(details) {
+        if ( details.content !== '' ) {
+            content = details.content;
+        }
+        nanoFetchResourceByKey('nano-resources', nanoOnExtendedResourcesLoaded);
+    };
 
     var onSelfieReady = function(success) {
         if ( success !== true ) {
-            µb.assets.get('ublock-resources', onResourcesLoaded);
+            nanoFetchResourceByKey('ublock-resources', nanoOnResourcesLoaded);
         }
     };
 
+    if ( nanoHasChange ) {
+        nanoFetchResourceByKey('ublock-resources', nanoOnResourcesLoaded);
+        return;
+    }
+
     µb.redirectEngine.resourcesFromSelfie(onSelfieReady);
 };
 
@@ -1237,7 +1264,10 @@
             }
         }
         // https://github.com/gorhill/uBlock/issues/2594
-        if ( details.assetKey === 'ublock-resources' ) {
+        if (
+                details.assetKey === 'ublock-resources' ||
+                details.assetKey === 'nano-resources'
+        ) {
             if (
                 this.hiddenSettings.ignoreRedirectFilters === true &&
                 this.hiddenSettings.ignoreScriptInjectFilters === true
@@ -1275,6 +1305,11 @@
             if ( cached ) {
                 this.loadRedirectResources(details.content);
             }
+        } else if ( details.assetKey === 'nano-resources' ) {
+            this.redirectEngine.invalidateResourcesSelfie();
+            if ( cached ) {
+                this.loadRedirectResources(details.content, true);
+            }
         }
         vAPI.messaging.broadcast({
             what: 'assetUpdated',
