diff --git a/src/js/assets.js b/src/js/assets.js
index 9317bba0..de28ff19 100644
--- a/src/js/assets.js
+++ b/src/js/assets.js
@@ -950,6 +950,9 @@ var updateNext = function() {
             if ( assetKey === 'ublock-resources' && noRemoteResources ) {
                 continue;
             }
+            if ( assetKey === 'nano-resources' && noRemoteResources ) {
+                continue;
+            }
             if (
                 fireNotification(
                     'before-asset-updated',
diff --git a/src/js/messaging.js b/src/js/messaging.js
index c0ec034c..0effde4f 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1073,10 +1073,14 @@ var onMessage = function(request, sender, callback) {
         µb.assets.purge(request.assetKey);
         µb.assets.remove('compiled/' + request.assetKey);
         // https://github.com/gorhill/uBlock/pull/2314#issuecomment-278716960
-        if ( request.assetKey === 'ublock-filters' ) {
+        if ( request.assetKey.startsWith('ublock-') ) {
             µb.assets.purge('ublock-resources');
             µb.redirectEngine.invalidateResourcesSelfie();
         }
+        if ( request.assetKey.startsWith('nano-') ) {
+            µb.assets.purge('nano-resources');
+            µb.redirectEngine.invalidateResourcesSelfie();
+        }
         break;
 
     case 'readHiddenSettings':
diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index 038079e1..ac43aece 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -194,7 +194,7 @@
             }
             content =
                 'try {\n' +
-                    content + '\n' +
+                    content.replace('{{nano}}', '') + '\n' +
                 '} catch ( e ) { }';
             scriptletCache.add(raw, content);
         }
@@ -256,6 +256,14 @@
             return;
         }
 
+        var nanoInjectArgs = parsed.suffix.slice(4, -1);
+        if (
+            !nano.cf.is_privileged &&
+            nanoInjectArgs.startsWith(nano.privileged_assets_previx)
+        ) {
+            return;
+        }
+
         // https://github.com/gorhill/uBlock/issues/3375
         //   Ignore instances of exception filter with negated hostnames,
         //   because there is no way to create an exception to an exception.
diff --git a/src/js/storage.js b/src/js/storage.js
index 4d307415..157c1986 100644
--- a/src/js/storage.js
+++ b/src/js/storage.js
@@ -994,7 +994,7 @@
 
 /******************************************************************************/
 
-µBlock.loadRedirectResources = function(updatedContent) {
+µBlock.loadRedirectResources = function(updatedContent, nanoIsExtended) {
     let content = '';
 
     const onDone = ( ) => {
@@ -1008,9 +1008,28 @@
         onDone();
     };
 
-    const onResourcesLoaded = details => {
+    const nanoHasChange =
+        typeof updatedContent === 'string' && updatedContent.length !== 0;
+
+    const nanoFetchResourceByKey = (key, callback) => {
+        if ( !nanoHasChange ) {
+            this.assets.get(key, callback);
+            return;
+        }
+        if ( key === 'ublock-resources' && !nanoIsExtended ) {
+            callback({ content: updatedContent });
+            return;
+        }
+        if ( key === 'nano-resources' && nanoIsExtended ) {
+            callback({ content: updatedContent });
+            return;
+        }
+        this.assets.get(key, callback);
+    };
+
+    const nanoOnExtendedResourcesLoaded = details => {
         if ( details.content !== '' ) {
-            content = details.content;
+            content += '\n\n' + details.content;
         }
         if ( this.hiddenSettings.userResourcesLocation === 'unset' ) {
             return onDone();
@@ -1021,13 +1040,21 @@
         );
     };
 
-    if ( typeof updatedContent === 'string' && updatedContent.length !== 0 ) {
-        return onResourcesLoaded({ content: updatedContent });
+    const nanoOnResourcesLoaded = details => {
+        if ( details.content !== '' ) {
+            content = details.content;
+        }
+        nanoFetchResourceByKey('nano-resources', nanoOnExtendedResourcesLoaded);
+    };
+
+    if ( nanoHasChange ) {
+        nanoFetchResourceByKey('ublock-resources', nanoOnResourcesLoaded);
+        return;
     }
 
     this.redirectEngine.resourcesFromSelfie().then(success => {
         if ( success !== true ) {
-            this.assets.get('ublock-resources', onResourcesLoaded);
+            nanoFetchResourceByKey('ublock-resources', nanoOnResourcesLoaded);
         }
     });
 };
@@ -1317,7 +1344,10 @@
             }
         }
         // https://github.com/gorhill/uBlock/issues/2594
-        if ( details.assetKey === 'ublock-resources' ) {
+        if (
+                details.assetKey === 'ublock-resources' ||
+                details.assetKey === 'nano-resources'
+        ) {
             if (
                 this.hiddenSettings.ignoreRedirectFilters === true &&
                 this.hiddenSettings.ignoreScriptInjectFilters === true
@@ -1360,6 +1390,11 @@
             if ( cached ) {
                 this.loadRedirectResources(details.content);
             }
+        } else if ( details.assetKey === 'nano-resources' ) {
+            this.redirectEngine.invalidateResourcesSelfie();
+            if ( cached ) {
+                this.loadRedirectResources(details.content, true);
+            }
         }
         vAPI.messaging.broadcast({
             what: 'assetUpdated',
