diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index fc40b388..16580dfb 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -177,6 +177,9 @@ const typeToMimeMap = new Map([
     [ 'xmlhttprequest', 'text/plain' ],
 ]);
 
+nano.insert_redirect_declaration(redirectableResources);
+nano.insert_mime_declaration(extToMimeMap);
+
 const validMimes = new Set(extToMimeMap.values());
 
 const mimeFromName = function(name) {
diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index 7c2b4890..f614e019 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -221,7 +221,7 @@
             }
             content =
                 'try {\n' +
-                    content + '\n' +
+                    content.replace('{{nano}}', '') + '\n' +
                 '} catch ( e ) { }';
             scriptletCache.add(rawToken, content);
         }
@@ -299,6 +299,14 @@
             return;
         }
 
+        var nanoInjectArgs = parsed.suffix.slice(4, -1);
+        if (
+            !nano.cf.is_privileged &&
+            nanoInjectArgs.startsWith(nano.privileged_assets_previx)
+        ) {
+            return;
+        }
+
         // https://github.com/gorhill/uBlock/issues/3375
         //   Ignore instances of exception filter with negated hostnames,
         //   because there is no way to create an exception to an exception.
