diff --git a/src/js/scriptlet-filtering.js b/src/js/scriptlet-filtering.js
index b1cd5141..5b5d606f 100644
--- a/src/js/scriptlet-filtering.js
+++ b/src/js/scriptlet-filtering.js
@@ -223,7 +223,7 @@
             }
             content =
                 'try {\n' +
-                    content + '\n' +
+                    content.replace('{{nano}}', '') + '\n' +
                 '} catch ( e ) { }';
             scriptletCache.add(rawToken, content);
         }
@@ -287,6 +287,14 @@
             return;
         }
 
+        var nanoInjectArgs = parsed.suffix.slice(4, -1);
+        if (
+            !nano.cf.is_privileged &&
+            nanoInjectArgs.startsWith(nano.privileged_assets_previx)
+        ) {
+            return;
+        }
+
         // https://github.com/gorhill/uBlock/issues/3375
         //   Ignore instances of exception filter with negated hostnames,
         //   because there is no way to create an exception to an exception.
