diff --git a/src/advanced-settings.html b/src/advanced-settings.html
index 4b260c58..54c4e8db 100644
--- a/src/advanced-settings.html
+++ b/src/advanced-settings.html
@@ -19,6 +19,7 @@
     <p><span data-i18n="advancedSettingsWarning"></span> <a class="fa info important" href="https://github.com/gorhill/uBlock/wiki/Advanced-settings" target="_blank">&#xf05a;</a>
     <p>
         <button id="advancedSettingsApply" class="custom important" type="button" disabled data-i18n="genericApplyChanges"></button>&ensp;
+        <button id="nanoForceRecompile" class="custom" type="button" data-i18n="nano_d_recompile_filters"></button>&ensp;
 </div>
 
 <div id="advancedSettings" class="codeMirrorContainer codeMirrorFillVertical"></div>
diff --git a/src/js/advanced-settings.js b/src/js/advanced-settings.js
index b65ad7a1..62d0a4ca 100644
--- a/src/js/advanced-settings.js
+++ b/src/js/advanced-settings.js
@@ -116,6 +116,17 @@ let applyChanges = function() {
 
 /******************************************************************************/
 
+let nanoForceRecompile = function() {
+    messaging.send(
+        'dashboard',
+        {
+            what: 'nanoForceRecompile'
+        }
+    );
+};
+
+/******************************************************************************/
+
 uDom.nodeFromId('advancedSettings').addEventListener(
     'input',
     advancedSettingsChanged
@@ -124,6 +135,10 @@ uDom.nodeFromId('advancedSettingsApply').addEventListener(
     'click',
     applyChanges
 );
+uDom.nodeFromId('nanoForceRecompile').addEventListener(
+    'click',
+    nanoForceRecompile
+);
 
 renderAdvancedSettings(true);
 
diff --git a/src/js/messaging.js b/src/js/messaging.js
index b8853689..f116ae1a 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1099,6 +1099,10 @@ var onMessage = function(request, sender, callback) {
         µb.changeHiddenSettings(µb.hiddenSettingsFromString(request.content));
         break;
 
+    case 'nanoForceRecompile':
+        nano.recompile_filters();
+        break;
+
     default:
         return vAPI.messaging.UNHANDLED;
     }
