diff --git a/src/advanced-settings.html b/src/advanced-settings.html
index 046b60f..d119058 100644
--- a/src/advanced-settings.html
+++ b/src/advanced-settings.html
@@ -15,6 +15,8 @@
     <p>
         <button id="advancedSettingsApply" class="custom important" type="button" disabled data-i18n="genericApplyChanges"></button>&ensp;
     <p><textarea id="advancedSettings" dir="auto" spellcheck="false"></textarea>
+    <p>
+        <button id="nanoForceRecompile" class="custom" type="button" data-i18n="nano_d_recompile_filters"></button>&ensp;
 </div>
 
 <script src="js/vapi.js"></script>
diff --git a/src/js/advanced-settings.js b/src/js/advanced-settings.js
index 9742187..fb4f7f5 100644
--- a/src/js/advanced-settings.js
+++ b/src/js/advanced-settings.js
@@ -103,9 +103,21 @@ var applyChanges = function() {
 
 /******************************************************************************/
 
+var nanoForceRecompile = function() {
+    messaging.send(
+        'dashboard',
+        {
+            what: 'nano_force_recompile'
+        }
+    );
+};
+
+/******************************************************************************/
+
 // Handle user interaction
 uDom('#advancedSettings').on('input', advancedSettingsChanged);
 uDom('#advancedSettingsApply').on('click', applyChanges);
+uDom('#nanoForceRecompile').on('click', nanoForceRecompile);
 
 renderAdvancedSettings();
 
diff --git a/src/js/messaging.js b/src/js/messaging.js
index 682d930..af4a966 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1039,6 +1039,10 @@ var onMessage = function(request, sender, callback) {
         µb.changeHiddenSettings(µb.hiddenSettingsFromString(request.content));
         break;
 
+    case 'nanoForceRecompile':
+        nano.recompile_filters();
+        break;
+
     default:
         return vAPI.messaging.UNHANDLED;
     }
