diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index cbfdd424..4037300e 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -543,6 +543,8 @@ RedirectEngine.prototype.supportedTypes = new Map([
     [ 'subdocument', 'sub_frame' ],
     [ 'xhr', 'xmlhttprequest' ],
     [ 'xmlhttprequest', 'xmlhttprequest' ],
+
+    [ 'iframe', 'sub_frame' ],
 ]);
 
 /******************************************************************************/
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 3de866df..bc511f72 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -161,6 +161,8 @@ const toNormalizedType = {
     'xmlhttprequest': 'xmlhttprequest',
             'webrtc': 'unsupported',
          'websocket': 'websocket',
+
+            'iframe': 'subdocument',
 };
 
 const typeValueFromCatBits = catBits => (catBits >>> 4) & 0b11111;
