diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index c9b176fd..c5afd71c 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -554,6 +554,8 @@ RedirectEngine.prototype.supportedTypes = new Map([
     [ 'subdocument', 'sub_frame' ],
     [ 'xhr', 'xmlhttprequest' ],
     [ 'xmlhttprequest', 'xmlhttprequest' ],
+
+    [ 'iframe', 'sub_frame' ],
 ]);
 
 /******************************************************************************/
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 4d92d478..7c7850cd 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1870,6 +1870,9 @@ FilterParser.prototype.toNormalizedType = {
     'xmlhttprequest': 'xmlhttprequest',
             'webrtc': 'unsupported',
          'websocket': 'websocket',
+
+             'ghide': 'generichide',
+            'iframe': 'subdocument',
 };
 
 /******************************************************************************/
