diff --git a/src/js/redirect-engine.js b/src/js/redirect-engine.js
index 6a938b35..412636e4 100644
--- a/src/js/redirect-engine.js
+++ b/src/js/redirect-engine.js
@@ -389,6 +389,8 @@ RedirectEngine.prototype.supportedTypes = new Map([
     [ 'subdocument', 'sub_frame' ],
     [ 'xhr', 'xmlhttprequest' ],
     [ 'xmlhttprequest', 'xmlhttprequest' ],
+
+    [ 'iframe', 'sub_frame' ],
 ]);
 
 /******************************************************************************/
diff --git a/src/js/static-net-filtering.js b/src/js/static-net-filtering.js
index 02baadbe..420cef0c 100644
--- a/src/js/static-net-filtering.js
+++ b/src/js/static-net-filtering.js
@@ -1847,6 +1847,9 @@ FilterParser.prototype.toNormalizedType = {
     'xmlhttprequest': 'xmlhttprequest',
             'webrtc': 'unsupported',
          'websocket': 'websocket',
+
+             'ghide': 'generichide',
+            'iframe': 'subdocument',
 };
 
 /******************************************************************************/
@@ -1938,7 +1941,9 @@ FilterParser.prototype.parseDomainOption = function(s) {
 
 FilterParser.prototype.parseOptions = function(s) {
     this.fopts = s;
-    for ( let opt of s.split(/\s*,\s*/) ) {
+    const opts = s.split(/\s*,\s*/);
+    for ( let i = 0; i < opts.length; i++ ) {
+        let opt = opts[i];
         const not = opt.startsWith('~');
         if ( not ) {
             opt = opt.slice(1);
@@ -1950,7 +1955,7 @@ FilterParser.prototype.parseOptions = function(s) {
         // https://issues.adblockplus.org/ticket/616
         // `generichide` concept already supported, just a matter of
         // adding support for the new keyword.
-        if ( opt === 'elemhide' || opt === 'generichide' ) {
+        if ( opt === 'generichide' || opt === 'ghide' || opt === 'elemhide' ) {
             if ( not === false ) {
                 this.parseTypeOption('generichide', false);
                 continue;
@@ -2006,8 +2011,11 @@ FilterParser.prototype.parseOptions = function(s) {
             this.dataStr = '';
             continue;
         }
-        // Used by Adguard, purpose is unclear -- just ignore for now.
-        if ( opt === 'empty' ) {
+        if ( opt === 'mp4' ) {
+            opts.push('media', 'redirect=noopmp4-1s');
+            this.fopts = opts.join(',');
+            var pos = this.raw.lastIndexOf('$');
+            this.raw = this.raw.slice(0, pos + 1) + this.fopts;
             continue;
         }
         // https://github.com/uBlockOrigin/uAssets/issues/192
