diff --git a/src/js/messaging.js b/src/js/messaging.js
index e1f34a0b..458cf602 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1176,6 +1176,13 @@ const onMessage = function(request, sender, callback) {
         nano.recompile_filters();
         break;
 
+    case 'nanoGetFilterLinterResult':
+        response = {
+            errors: nano.fl.errors,
+            warnings: nano.fl.warnings
+        };
+        break;
+
     default:
         return vAPI.messaging.UNHANDLED;
     }
