diff --git a/src/js/messaging.js b/src/js/messaging.js
index 576167da..7a040de1 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1184,6 +1184,13 @@ const onMessage = function(request, sender, callback) {
         nano.recompile_filters();
         break;
 
+    case 'nanoGetFilterLinterResult':
+        response = {
+            errors: nano.fl.errors,
+            warnings: nano.fl.warnings
+        };
+        break;
+
     default:
         return vAPI.messaging.UNHANDLED;
     }
