diff --git a/src/js/messaging.js b/src/js/messaging.js
index d3f50ee3..eb72003f 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -1108,6 +1108,13 @@ var onMessage = function(request, sender, callback) {
         nano.recompile_filters();
         break;
 
+    case 'nanoGetFilterLinterResult':
+        response = {
+            errors: nano.fl.errors,
+            warnings: nano.fl.warnings
+        };
+        break;
+
     default:
         return vAPI.messaging.UNHANDLED;
     }
