diff --git a/src/css/popup.css b/src/css/popup.css
index 6a1c5f7f..9d2b5e06 100644
--- a/src/css/popup.css
+++ b/src/css/popup.css
@@ -16,7 +16,6 @@ h2 {
     background-color: #eee;
     border: 0;
     color: #666;
-    cursor: pointer;
     font-size: 100%;
     font-weight: normal;
     padding: 0.2em;
@@ -98,18 +97,22 @@ p {
 #switch .fa-icon {
     fill: #0046ff;
     cursor: pointer;
-    font-size: 700%;
+    font-size: 600%;
     margin: 0;
     padding: 0;
     }
 #switch .fa-icon:hover {
-    opacity: 0.9;
+    opacity: 0.8;
     }
 body.off #switch .fa-icon {
     fill: #ccc;
     }
-#basicTools {
+#basicTools, #basicTools2 {
     margin: 1.2em 0 0.5em 0;
+    text-align: center;
+    }
+#basicTools {
+    margin-bottom: 1.2em;
     }
 .tool:first-child {
     margin-left: 0;
@@ -181,26 +184,22 @@ body.responsive #extraTools {
     fill: #222;
     }
 
+#nanoCacheControlStrip {
+    margin: 0px;
+    }
 #refresh {
-    background-color: #ffe;
-    border: 1px solid #ddc;
-    border-radius: 4px;
     cursor: pointer;
-    display: none;
+    display: flex;
     fill: #888;
-    font-size: 3em;
+    font-size: 2.3em;
     justify-content: center;
-    left: 4px;
     line-height: 1;
-    padding: 0.15em;
-    position: absolute;
-    right: 4px;
     }
 body.dirty #refresh {
-    display: flex;
+    fill: red;
     }
-body.dirty #refresh:hover {
-    fill: #000
+#refresh:hover {
+    fill: #444 !important;
     }
 
 #tooltip {
@@ -230,8 +229,6 @@ body[dir="rtl"] #tooltip {
     direction: rtl;
     }
 #tooltip.show {
-    transition: opacity 0.15s 0.5s;
-    -webkit-transition: opacity 0.15s 0.5s;
     visibility: visible;
     opacity: 1;
     }
diff --git a/src/js/popup.js b/src/js/popup.js
index b2c78b01..25b0a432 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -552,7 +552,7 @@ const renderTooltips = function(selector) {
 
 const tooltipTargetSelectors = new Map([
     [
-        '#switch',
+        '#switch .fa-icon',
         {
             state: 'body.off',
             i18n: 'popupPowerSwitchInfo',
@@ -711,7 +711,7 @@ const toggleNetFilteringSwitch = function(ev) {
             tabId: popupData.tabId
         }
     );
-    renderTooltips('#switch');
+    renderTooltips('#switch .fa-icon');
     hashFromPopupData();
 };
 
@@ -1134,10 +1134,10 @@ const onHideTooltip = function() {
     getPopupData(tabId);
 })();
 
-uDom('#switch').on('click', toggleNetFilteringSwitch);
+uDom('#switch .fa-icon').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
-uDom('h2').on('click', toggleFirewallPane);
+uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('#refresh').on('click', reloadTab);
 uDom('.hnSwitch').on('click', toggleHostnameSwitch);
 uDom('#saveRules').on('click', saveFirewallRules);
diff --git a/src/popup.html b/src/popup.html
index 57366f74..97d8b974 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -15,23 +15,29 @@
 <div id="appinfo"><span id="appname">&nbsp;</span>&nbsp;<span id="version">&nbsp;</span></div>
 <div id="panes">
     <div class="tooltipContainer">
-        <p id="switch" role="button" aria-label data-tip-position="under" tabindex="0"><span class="fa-icon">power-off</span></p>
+        <p id="switch" role="button" tabindex="0"><span class="fa-icon" aria-label data-tip-position="under">power-off</span></p>
         <p id="basicTools">
-            <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
             <span id="gotoPick" class="fa-icon tool" data-i18n-tip="popupTipPicker" data-tip-position="under">eye-dropper</span>
+            <span id="nanoFirewallPaneToggle" class="fa tool enabled" data-i18n-tip="nano_p_open_firewall_pane" data-tip-position="under">&#xf0db;</span>
             <a href="logger-ui.html#_" class="fa-icon tool enabled" aria-label="data-tip" data-i18n-tip="popupTipLog" data-tip-position="under" target="uBOLogger" tabindex="0">list-alt</a>
-            <a href="dashboard.html" class="fa-icon tool enabled" aria-label="data-tip" data-i18n-tip="popupTipDashboard" data-tip-position="under" target="uBODashboard" tabindex="0">sliders</a>
+            <a href="dashboard.html" class="fa tool enabled" aria-label="data-tip" data-i18n-tip="popupTipDashboard" data-tip-position="under" target="uBODashboard" tabindex="0">&#xf0ad;</a>
+        </p>
+        <p id="basicTools2">
+            <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
             <span data-i18n="popupBlockedOnThisPagePrompt">&nbsp;</span>
         </p>
         <p class="statValue" id="page-blocked">?</p>
-        <div id="refresh" class="fa-icon">refresh</div>
         <p class="statName">
             <span data-i18n="popupBlockedSinceInstallPrompt">&nbsp;</span>
         </p>
         <p class="statValue" id="total-blocked">?</p>
+        <h2 data-i18n="nano_p_cache_control">&nbsp;</h2>
+        <p id="nanoCacheControlStrip">
+            <span id="refresh" class="fa-icon" data-i18n-tip="nano_p_refresh_button">refresh</span>
+        </p>
         <h2 data-i18n="popupHitDomainCountPrompt">&nbsp;</h2>
         <p class="statValue" id="popupHitDomainCount">&nbsp;</p>
         <div id="extraTools">
