diff --git a/src/js/messaging.js b/src/js/messaging.js
index 99b4720f..3acd1d88 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -140,6 +140,10 @@ var onMessage = function(request, sender, callback) {
         µb.elementPickerExec(request.tabId, request.targetURL, request.zap);
         break;
 
+    case 'nanoInjectForceScrollCSS':
+        nano.inject_force_scroll(request.tabId);
+        break;
+
     case 'gotoURL':
         µb.openNewTab(request.details);
         break;
diff --git a/src/js/popup.js b/src/js/popup.js
index 18acc2be..90a62898 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -432,6 +432,7 @@ const renderPopup = function() {
                            popupData.netFilteringSwitch === true;
     uDom.nodeFromId('gotoPick').classList.toggle('enabled', canElementPicker);
     uDom.nodeFromId('gotoZap').classList.toggle('enabled', canElementPicker);
+    uDom.nodeFromId('nanoForceEnableScroll').classList.toggle('enabled', canElementPicker);
 
     let blocked = popupData.pageBlockedRequestCount,
         total = popupData.pageAllowedRequestCount + blocked,
@@ -729,6 +730,20 @@ const gotoPick = function() {
 
 /******************************************************************************/
 
+const nanoForceEnableScroll = function() {
+    messaging.send(
+        'popupPanel',
+        {
+            what: 'nanoInjectForceScrollCSS',
+            tabId: popupData.tabId
+        }
+    );
+
+    vAPI.closePopup();
+};
+
+/******************************************************************************/
+
 const gotoURL = function(ev) {
     if ( this.hasAttribute('href') === false ) { return; }
 
@@ -1119,6 +1134,7 @@ const onHideTooltip = function() {
 uDom('#switch .fa').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
+uDom('#nanoForceEnableScroll').on('click', nanoForceEnableScroll);
 uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('#refresh').on('click', reloadTab);
 uDom('.hnSwitch').on('click', toggleHostnameSwitch);
diff --git a/src/popup.html b/src/popup.html
index a38ad3ab..4cb4905b 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -24,6 +24,7 @@
         </p>
         <p id="basicTools2">
             <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
+            <span id="nanoForceEnableScroll" class="fa-icon tool" data-i18n-tip="nano_p_force_scroll" data-tip-position="under">bolt</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
