diff --git a/src/js/messaging.js b/src/js/messaging.js
index 6e5c956a..41bcd4a1 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -137,6 +137,10 @@ const onMessage = function(request, sender, callback) {
         µb.elementPickerExec(request.tabId, request.targetURL, request.zap);
         break;
 
+    case 'nanoInjectForceScrollCSS':
+        nano.inject_force_scroll(request.tabId);
+        break;
+
     case 'gotoURL':
         µb.openNewTab(request.details);
         break;
diff --git a/src/js/popup.js b/src/js/popup.js
index c7290995..05956a64 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -462,6 +462,7 @@ const renderPopup = function() {
                            popupData.netFilteringSwitch === true;
     uDom.nodeFromId('gotoPick').classList.toggle('enabled', canElementPicker);
     uDom.nodeFromId('gotoZap').classList.toggle('enabled', canElementPicker);
+    uDom.nodeFromId('nanoForceEnableScroll').classList.toggle('enabled', canElementPicker);
 
     let blocked = popupData.pageBlockedRequestCount,
         total = popupData.pageAllowedRequestCount + blocked,
@@ -767,6 +768,20 @@ const gotoPick = function() {
 
 /******************************************************************************/
 
+const nanoForceEnableScroll = function() {
+    messaging.send(
+        'popupPanel',
+        {
+            what: 'nanoInjectForceScrollCSS',
+            tabId: popupData.tabId
+        }
+    );
+
+    vAPI.closePopup();
+};
+
+/******************************************************************************/
+
 const gotoURL = function(ev) {
     if ( this.hasAttribute('href') === false ) { return; }
 
@@ -1204,6 +1219,7 @@ const onHideTooltip = function() {
 uDom('#switch .fa-icon').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
+uDom('#nanoForceEnableScroll').on('click', nanoForceEnableScroll);
 uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('.hnSwitch').on('click', ev => { toggleHostnameSwitch(ev); });
 uDom('#saveRules').on('click', saveFirewallRules);
diff --git a/src/popup.html b/src/popup.html
index 282de88d..ddcf72bf 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -23,6 +23,7 @@
         </p>
         <p id="basicTools2">
             <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
+            <span id="nanoForceEnableScroll" class="fa tool" data-i18n-tip="nano_p_force_scroll" data-tip-position="under">&#xf103;</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
