diff --git a/src/js/messaging.js b/src/js/messaging.js
index 70b7eb12..aa5a102d 100644
--- a/src/js/messaging.js
+++ b/src/js/messaging.js
@@ -140,6 +140,10 @@ var onMessage = function(request, sender, callback) {
         µb.elementPickerExec(request.tabId, request.targetURL, request.zap);
         break;
 
+    case 'nanoInjectForceScrollCSS':
+        nano.inject_force_scroll(request.tabId);
+        break;
+
     case 'gotoURL':
         µb.openNewTab(request.details);
         break;
diff --git a/src/js/popup.js b/src/js/popup.js
index 0ff2bc23..092ecdcf 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -436,6 +436,7 @@ var renderPopup = function() {
                            popupData.netFilteringSwitch === true;
     uDom.nodeFromId('gotoPick').classList.toggle('enabled', canElementPicker);
     uDom.nodeFromId('gotoZap').classList.toggle('enabled', canElementPicker);
+    uDom.nodeFromId('nanoForceEnableScroll').classList.toggle('enabled', canElementPicker);
 
     let blocked = popupData.pageBlockedRequestCount,
         total = popupData.pageAllowedRequestCount + blocked,
@@ -733,6 +734,20 @@ var gotoPick = function() {
 
 /******************************************************************************/
 
+var nanoForceEnableScroll = function() {
+    messaging.send(
+        'popupPanel',
+        {
+            what: 'nanoInjectForceScrollCSS',
+            tabId: popupData.tabId
+        }
+    );
+
+    vAPI.closePopup();
+};
+
+/******************************************************************************/
+
 var gotoURL = function(ev) {
     if ( this.hasAttribute('href') === false ) { return; }
 
@@ -1128,6 +1143,7 @@ var onHideTooltip = function() {
 uDom('#switch .fa').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
+uDom('#nanoForceEnableScroll').on('click', nanoForceEnableScroll);
 uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('#refresh').on('click', reloadTab);
 uDom('.hnSwitch').on('click', toggleHostnameSwitch);
diff --git a/src/popup.html b/src/popup.html
index d0aec45f..9297c527 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -23,6 +23,7 @@
         </p>
         <p id="basicTools2">
             <span id="gotoZap" class="fa tool" data-i18n-tip="popupTipZapper" data-tip-position="under">&#xf0e7;</span>
+            <span id="nanoForceEnableScroll" class="fa tool" data-i18n-tip="nano_p_force_scroll" data-tip-position="under">&#xf103;</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
