diff --git a/platform/chromium/vapi-common.js b/platform/chromium/vapi-common.js
index 0454086d..2b58510d 100644
--- a/platform/chromium/vapi-common.js
+++ b/platform/chromium/vapi-common.js
@@ -135,6 +135,23 @@ vAPI.download = function(details) {
 
 vAPI.getURL = chrome.runtime.getURL;
 
+vAPI.nanoOpenBugReporter = function() {
+    const reporter = vAPI.getURL('/reporter/index.html');
+
+    chrome.tabs.query({
+        active: true,
+        currentWindow: true
+    }, function(tabs) {
+        if (chrome.runtime.lastError || tabs.length === 0) {
+            return;
+        }
+
+        chrome.tabs.create({
+            url: reporter + '?' + tabs[0].id
+        });
+    });
+};
+
 /******************************************************************************/
 
 vAPI.i18n = chrome.i18n.getMessage;
diff --git a/src/js/popup.js b/src/js/popup.js
index c982921a..b3ba1610 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -759,6 +759,8 @@ const nanoForceEnableScroll = function() {
     vAPI.closePopup();
 };
 
+const nanoOpenBugReporter = vAPI.nanoOpenBugReporter;
+
 /******************************************************************************/
 
 const gotoURL = function(ev) {
@@ -1153,6 +1155,7 @@ uDom('#switch .fa-icon').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
 uDom('#nanoForceEnableScroll').on('click', nanoForceEnableScroll);
+uDom('#nanoOpenBugReporter').on('click', nanoOpenBugReporter);
 uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('#refresh').on('click', reloadTab);
 uDom('.hnSwitch').on('click', toggleHostnameSwitch);
diff --git a/src/popup.html b/src/popup.html
index 48f78ec3..ab45a351 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -25,6 +25,7 @@
         <p id="basicTools2">
             <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
             <span id="nanoForceEnableScroll" class="fa tool" data-i18n-tip="nano_p_force_scroll" data-tip-position="under">&#xf103;</span>
+            <span id="nanoOpenBugReporter" class="fa tool enabled" data-i18n-tip="nano_p_bug_reporter" data-tip-position="under">&#xf088;</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
