diff --git a/src/js/popup.js b/src/js/popup.js
index feea0361..e98927e6 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -776,6 +776,10 @@ const nanoForceEnableScroll = function() {
     vAPI.closePopup();
 };
 
+const nanoOpenBugReporter = function() {
+    vAPI.nanoOpenBugReporter(popupData.tabId);
+};
+
 /******************************************************************************/
 
 const gotoURL = function(ev) {
@@ -1222,6 +1226,7 @@ uDom('#switch .fa-icon').on('click', toggleNetFilteringSwitch);
 uDom('#gotoZap').on('click', gotoZap);
 uDom('#gotoPick').on('click', gotoPick);
 uDom('#nanoForceEnableScroll').on('click', nanoForceEnableScroll);
+uDom('#nanoOpenBugReporter').on('click', nanoOpenBugReporter);
 uDom('#nanoFirewallPaneToggle').on('click', toggleFirewallPane);
 uDom('.hnSwitch').on('click', ev => { toggleHostnameSwitch(ev); });
 uDom('#saveRules').on('click', saveFirewallRules);
diff --git a/src/popup.html b/src/popup.html
index 9314430a..dd46b5ac 100644
--- a/src/popup.html
+++ b/src/popup.html
@@ -25,6 +25,7 @@
         <p id="basicTools2">
             <span id="gotoZap" class="fa-icon tool" data-i18n-tip="popupTipZapper" data-tip-position="under">bolt</span>
             <span id="nanoForceEnableScroll" class="fa tool" data-i18n-tip="nano_p_force_scroll" data-tip-position="under">&#xf103;</span>
+            <span id="nanoOpenBugReporter" class="fa tool enabled" data-i18n-tip="nano_p_bug_reporter" data-tip-position="under">&#xf088;</span>
         </p>
         <h2 id="dfToggler" data-i18n="popupBlockedRequestPrompt">&nbsp;</h2>
         <p class="statName">
