diff --git a/src/js/popup.js b/src/js/popup.js
index 94c9903c..6a6191ad 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -1120,4 +1120,10 @@ uDom('a[href]').on('click', gotoURL);
 
 /******************************************************************************/
 
+if ( typeof elib !== 'undefined' && !location.search ) {
+    elib.unbreak_popup('.fa { vertical-align: middle; }');
+}
+
+/******************************************************************************/
+
 })();
diff --git a/src/js/settings.js b/src/js/settings.js
index c7e24297..04a4c251 100644
--- a/src/js/settings.js
+++ b/src/js/settings.js
@@ -186,6 +186,15 @@ var synchronizeDOM = function() {
         'advancedUser',
         uDom.nodeFromId('advanced-user-enabled').checked === true
     );
+
+    if ( typeof elib !== 'undefined' ) {
+        var elem = document.querySelector('a[href="advanced-settings.html"]');
+        if ( document.body.classList.contains('advancedUser') ) {
+            elem.style.display = 'inline';
+        } else {
+            elem.style.display = 'none';
+        }
+    }
 };
 
 /******************************************************************************/
