diff --git a/src/js/popup.js b/src/js/popup.js
index b3ba1610..6adaa9e1 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -1170,4 +1170,10 @@ uDom('a[href]').on('click', gotoURL);
 
 /******************************************************************************/
 
+if ( typeof elib !== 'undefined' && !location.search ) {
+    elib.unbreak_popup('.fa { vertical-align: middle; }');
+}
+
+/******************************************************************************/
+
 })();
diff --git a/src/js/settings.js b/src/js/settings.js
index adc38102..5f83df03 100644
--- a/src/js/settings.js
+++ b/src/js/settings.js
@@ -184,6 +184,15 @@ const synchronizeDOM = function() {
         'advancedUser',
         uDom.nodeFromId('advanced-user-enabled').checked === true
     );
+
+    if ( typeof elib !== 'undefined' ) {
+        var elem = document.querySelector('a[href="advanced-settings.html"]');
+        if ( document.body.classList.contains('advancedUser') ) {
+            elem.style.display = 'inline';
+        } else {
+            elem.style.display = 'none';
+        }
+    }
 };
 
 /******************************************************************************/
diff --git a/src/js/ublock.js b/src/js/ublock.js
index 4afad498..f944edbe 100644
--- a/src/js/ublock.js
+++ b/src/js/ublock.js
@@ -429,7 +429,9 @@ var matchBucket = function(url, hostname, bucket, start) {
             runAt: 'document_end'
         },
         ( ) => {
-            vAPI.tabs.select(tabId);
+            if ( typeof elib === 'undefined' ) {
+                vAPI.tabs.select(tabId);
+            }
         }
     );
 };
