diff --git a/src/js/popup.js b/src/js/popup.js
index d05dde9a..c1c4ebbb 100644
--- a/src/js/popup.js
+++ b/src/js/popup.js
@@ -1235,4 +1235,10 @@ uDom('a[href]').on('click', gotoURL);
 
 /******************************************************************************/
 
+if ( typeof elib !== 'undefined' && !location.search ) {
+    elib.unbreak_popup('.fa { vertical-align: middle; }');
+}
+
+/******************************************************************************/
+
 })();
diff --git a/src/js/settings.js b/src/js/settings.js
index 9ba49559..01742b11 100644
--- a/src/js/settings.js
+++ b/src/js/settings.js
@@ -181,6 +181,15 @@ const synchronizeDOM = function() {
         'advancedUser',
         uDom.nodeFromId('advanced-user-enabled').checked === true
     );
+
+    if ( typeof elib !== 'undefined' ) {
+        var elem = document.querySelector('a[href="advanced-settings.html"]');
+        if ( document.body.classList.contains('advancedUser') ) {
+            elem.style.display = 'inline';
+        } else {
+            elem.style.display = 'none';
+        }
+    }
 };
 
 /******************************************************************************/
diff --git a/src/js/ublock.js b/src/js/ublock.js
index 6dd2a792..3120f096 100644
--- a/src/js/ublock.js
+++ b/src/js/ublock.js
@@ -435,7 +435,9 @@ const matchBucket = function(url, hostname, bucket, start) {
     // https://github.com/uBlockOrigin/uBlock-issues/issues/168
     //   Force activate the target tab once the element picker has been
     //   injected.
-    vAPI.tabs.select(tabId);
+    if ( typeof elib === 'undefined' ) {
+        vAPI.tabs.select(tabId);
+    }
 };
 
 /******************************************************************************/
