diff --git a/platform/chromium/vapi-background.js b/platform/chromium/vapi-background.js
index 18c1c631..121b75e3 100644
--- a/platform/chromium/vapi-background.js
+++ b/platform/chromium/vapi-background.js
@@ -61,24 +61,12 @@ vAPI.app = {
     version: manifest.version,
 
     intFromVersion: function(s) {
-        const parts = s.match(/(?:^|\.|b|rc)\d+/g);
-        if ( parts === null ) { return 0; }
-        let vint = 0;
-        for ( let i = 0; i < 4; i++ ) {
-            const pstr = parts[i] || '';
-            let pint;
-            if ( pstr === '' ) {
-                pint = 0;
-            } else if ( pstr.startsWith('.') || pstr.startsWith('b') ) {
-                pint = parseInt(pstr.slice(1), 10);
-            } else if ( pstr.startsWith('rc') ) {
-                pint = parseInt(pstr.slice(2), 10) + 100;
-            } else {
-                pint = parseInt(pstr, 10);
-            }
-            vint = vint * 1000 + pint;
+        const parts = s.split('.');
+        if ( parts[3] === '0' ) { // Legacy Edge
+            return parseInt(parts[2], 10);
+        } else {
+            return parseInt(parts[3], 10);
         }
-        return vint;
     },
 
     restart: function() {
diff --git a/src/js/start.js b/src/js/start.js
index d623bfe0..2bc98391 100644
--- a/src/js/start.js
+++ b/src/js/start.js
@@ -119,7 +119,7 @@ const onVersionReady = function(lastVersion) {
 
     // https://github.com/uBlockOrigin/uBlock-issues/issues/494
     //   Remove useless per-site switches.
-    if ( lastVersionInt <= 1019003007 ) {
+    if ( lastVersionInt <= 111 ) {
         µb.sessionSwitches.toggle('no-scripting', 'behind-the-scene', 0);
         µb.permanentSwitches.toggle('no-scripting', 'behind-the-scene', 0);
         µb.saveHostnameSwitches();
