diff --git a/src/js/start.js b/src/js/start.js
index e6099f4..220da63 100644
--- a/src/js/start.js
+++ b/src/js/start.js
@@ -129,47 +129,6 @@ var onVersionReady = function(lastVersion) {
         µb.redirectEngine.invalidateResourcesSelfie();
     }
 
-    // From 1.15.19b9 and above, the `behind-the-scene` scope is no longer
-    // whitelisted by default, and network requests from that scope will be
-    // subject to filtering by default.
-    //
-    // Following code is to remove the `behind-the-scene` scope when updating
-    // from a version older than 1.15.19b9.
-    // This will apply only to webext versions of uBO, as the following would
-    // certainly cause too much breakage in Firefox legacy given that uBO can
-    // see ALL network requests.
-    // Remove when everybody is beyond 1.15.19b8.
-    (function patch1015019008(s) {
-        if ( vAPI.firefox !== undefined ) { return; }
-        var match = /^(\d+)\.(\d+)\.(\d+)(?:\D+(\d+))?/.exec(s);
-        if ( match === null ) { return; }
-        var v =
-            parseInt(match[1], 10) * 1000 * 1000 * 1000 +
-            parseInt(match[2], 10) * 1000 * 1000 +
-            parseInt(match[3], 10) * 1000 +
-            (match[4] ? parseInt(match[4], 10) : 0);
-        if ( /rc\d+$/.test(s) ) { v += 100; }
-        if ( v > 1015019008 ) { return; }
-        if ( µb.getNetFilteringSwitch('http://behind-the-scene/') ) { return; }
-        var fwRules = [
-            'behind-the-scene * * noop',
-            'behind-the-scene * image noop',
-            'behind-the-scene * 3p noop',
-            'behind-the-scene * inline-script noop',
-            'behind-the-scene * 1p-script noop',
-            'behind-the-scene * 3p-script noop',
-            'behind-the-scene * 3p-frame noop'
-        ].join('\n');
-        µb.sessionFirewall.fromString(fwRules, true);
-        µb.permanentFirewall.fromString(fwRules, true);
-        µb.savePermanentFirewallRules();
-        µb.hnSwitches.fromString([
-            'no-large-media: behind-the-scene false'
-        ].join('\n'), true);
-        µb.saveHostnameSwitches();
-        µb.toggleNetFilteringSwitch('http://behind-the-scene/', '', true);
-    })(lastVersion);
-
     vAPI.storage.set({ version: vAPI.app.version });
 };
 
